### PR TITLE
Fix min(_, n) /max(_, n) ordering with NaN's in input

### PR DESCRIPTION
**Pull request by kgpai**
**What**
We add custom comparison functors in place of the default ones for Presto max(_, n)/min(_,n) aggregate functions. 

**Why**
As per #8749 with +/- NaN behavior of max/min is different from what is seen in PrestoJava. Presto Java seems to treat NaN > infinity whereas c++ doesnt and  moves to end of list if its +Nan and moves to beginning of list if its -NaN. 

**Notes**
N/A

### CODE REVIEW

File: velox/functions/prestosql/aggregates/MinMaxAggregates.cpp [hunk 1]
```diff
@@ -886,18 +886,56 @@ class MinMaxNAggregateBase : public exec::Aggregate {
...
+/// We have to use custom comparators to ensure compatibility with Presto Java.
+/// Currently Presto Java treats (+/- Nan) > Infinity, however C++ treats NaN's
+/// as nulls and moves them to end of the list if its +NaN and beginning of list
+/// if its -NaN.
 template <typename T>
-class MinNAggregate : public MinMaxNAggregateBase<T, std::less<T>> {
+struct compareLess {
+  bool operator()(const T& lhs, const T& rhs) {
+    if constexpr (std::is_same_v<T, double> || std::is_same_v<T, float>) {
+      if (std::isnan(lhs)) {
```
Comment (Yuhta) [inline velox/functions/prestosql/aggregates/MinMaxAggregates.cpp:897]: If we decide to go this route, we need to do it more efficiently and systematically.  The following two functions should be kept separately (e.g. in `PrestoComparisonUtil`) and used wherever applies:
1. First we need to normalize all `-NaN` to `NaN` in input with a global function
2. Optimize the comparison
```c++
if constexpr (std::is_same_v<T, double>) {
  return reinterpret_cast<int64_t>(lhs) < reinterpret_cast<int64_t>(rhs);
}
if constexpr (std::is_same_v<T, float>) {
  return reinterpret_cast<int32_t>(lhs) < reinterpret_cast<int32_t>(rhs);
}
```

**Review by mbasmanova — CHANGES_REQUESTED** [review]
> Presto Java seems to treat NaN > infinity

I don't see this being the case.

```
presto:di> select max(x), min(x) from unnest(array[1.0, nan(), infinity()]) as t(x);
  _col0   | _col1
----------+-------
 Infinity |   1.0
(1 row)
```

**Comment by kgpai** [comment]
@mbasmanova These changes are realted to min(_, n) and max(_, n) signatures. 

``` 
select max(x, 5), min(x, 5) from (values Nan(), INFINITY(), 1, -NAN(), 2) as T(x) ; 

--  [   "NaN",   "NaN",   "Infinity",   2,   1 ] | [   1,   2,   "Infinity",   "NaN",   "NaN" ]
```

File: velox/functions/prestosql/aggregates/MinMaxAggregates.cpp (hunk 1 above)
Comment (kgpai) [reply velox/functions/prestosql/aggregates/MinMaxAggregates.cpp:897]: I can move these to a common util class, one thing I am not so sure is reinterpreting doubles as int64_t , is that always correct ? Does it work no matter the values ?

**Comment by mbasmanova** [comment]
@kgpai Thank you for clarifying. It looks like min and min_n are not consistent with each other and min(x) != min(x, 1). That's bad. I'm not sure if it makes sense for Velox to match that.

```
presto:di> select max(x), min(x) from unnest(array[1.0, nan(), infinity()]) as t(x);
  _col0   | _col1
----------+-------
 Infinity |   1.0
(1 row)

Query 20240215_212745_20524_c3458, FINISHED, 1 node
Splits: 6 total, 6 done (100.00%)
[Latency: client-side: 156ms, server-side: 74ms] [0 rows, 0B] [0 rows/s, 0B/s]

presto:di> select max(x, 1), min(x, 1) from unnest(array[1.0, nan(), infinity()]) as t(x);
 _col0 | _col1
-------+-------
 [NaN] | [1.0]
(1 row)
```

File: velox/functions/prestosql/aggregates/MinMaxAggregates.cpp (hunk 1 above)
Comment (Yuhta) [reply velox/functions/prestosql/aggregates/MinMaxAggregates.cpp:897]: Yes it works for everything except -NaN values, which we will normalize out

**Comment by kgpai** [comment]
Comment to keep PR alive

**Pull request closed without merging**